### PR TITLE
Add possibility to invent seeding stubs coordinates before DR

### PR DESCRIPTION
**Pull request by sarafiorendi**
PR description:
Include code to enable the setting of seeding stubs coordinates to those inferred from the tracklet trajectory, before the DR step.
Performance shown [here](https://indico.cern.ch/event/1170133/#51-invented-seed-stubs-from-tr)

### CODE REVIEW

File: L1Trigger/TrackFindingTracklet/interface/PurgeDuplicate.h [hunk 1]
```diff
@@ -37,6 +37,12 @@ namespace trklet {
 
   private:
     double getPhiRes(Tracklet* curTracklet, const Stub* curStub);
+    bool isSeedingStub(int, int, int);
+    std::string l1tinfo(const L1TStub*, std::string);
+    std::pair<int, int> findLayerDisk(const Stub*);
+    std::vector<double> getInventedCoords(unsigned int, const Stub*, Tracklet*);
```
Comment (tomalin) [inline L1Trigger/TrackFindingTracklet/interface/PurgeDuplicate.h]: Add comments explaining what these two "invented" functions do.

File: L1Trigger/TrackFindingTracklet/interface/PurgeDuplicate.h [hunk 2]
```diff
@@ -37,6 +37,12 @@ namespace trklet {
...
+    std::vector<double> getInventedCoordsExtended(unsigned int, const Stub*, Tracklet*);
+    std::vector<const Stub*> seedStubCoordsFromTracklet(unsigned int, Tracklet*, std::vector<const Stub*>);
```
Comment (tomalin) [inline L1Trigger/TrackFindingTracklet/interface/PurgeDuplicate.h]: Add comment explaining what seedStubCoordsFromTracklet does. Also the name of this function is misleading. It doesn't return the coordinates. It returns the invented stubs.

File: L1Trigger/TrackFindingTracklet/src/PurgeDuplicate.cc [hunk 3]
```diff
@@ -505,13 +534,165 @@ double PurgeDuplicate::getPhiRes(Tracklet* curTracklet, const Stub* curStub) {
...
+        inv_r_z_phi = getInventedCoordsExtended(iSector, thisStub, tracklet);
+      }
+      double stub_x_invent = inv_r_z_phi[0] * std::cos(inv_r_z_phi[2]);
+      double stub_y_invent = inv_r_z_phi[0] * std::sin(inv_r_z_phi[2]);
+      double stub_z_invent = inv_r_z_phi[1];
+
+      Stub* invent_stub_ptr = new Stub(*thisStub);
+      const L1TStub* L1stub = thisStub->l1tstub();
+
+      L1TStub invent_L1stub(L1stub->DTClink(),
```
Comment (tomalin) [inline L1Trigger/TrackFindingTracklet/src/PurgeDuplicate.cc]: This seems a long-winded way to copy an L1TStub. Can't you just profit from L1TStub's default copy constructor and do:
L1TStub invent_L1TStub = L1stub;
invent_L1TStub.setCoords(x,y,z);
where you also need to add function L1TStub::setCoords(x,y,z);

File: L1Trigger/TrackFindingTracklet/src/PurgeDuplicate.cc [hunk 4]
```diff
@@ -505,13 +534,165 @@ double PurgeDuplicate::getPhiRes(Tracklet* curTracklet, const Stub* curStub) {
...
+        inv_r_z_phi = getInventedCoords(iSector, thisStub, tracklet);
+      else {
+        inv_r_z_phi = getInventedCoordsExtended(iSector, thisStub, tracklet);
+      }
+      double stub_x_invent = inv_r_z_phi[0] * std::cos(inv_r_z_phi[2]);
+      double stub_y_invent = inv_r_z_phi[0] * std::sin(inv_r_z_phi[2]);
+      double stub_z_invent = inv_r_z_phi[1];
+
+      Stub* invent_stub_ptr = new Stub(*thisStub);
+      const L1TStub* L1stub = thisStub->l1tstub();
```
Comment (tomalin) [inline L1Trigger/TrackFindingTracklet/src/PurgeDuplicate.cc]: All variable names in CMSSW must start with a lower case letter.

File: L1Trigger/TrackFindingTracklet/python/Tracklet_cfi.py [hunk 5]
```diff
@@ -13,6 +13,7 @@
                                                asciiFileName = cms.untracked.string(""),
                                                Extended = cms.bool(False),
                                                Reduced = cms.bool(False),
+                                               InventStubs = cms.bool(False),
```
Comment (tomalin) [inline L1Trigger/TrackFindingTracklet/python/Tracklet_cfi.py]: Is there a reason this parameter needs to be in Tracklet_cfi.py, and not just specified in Settings.h?

File: L1Trigger/TrackFindingTracklet/interface/PurgeDuplicate.h (hunk 1 above)
Comment (sarafiorendi) [reply L1Trigger/TrackFindingTracklet/interface/PurgeDuplicate.h]: done

File: L1Trigger/TrackFindingTracklet/interface/PurgeDuplicate.h (hunk 2 above)
Comment (sarafiorendi) [reply L1Trigger/TrackFindingTracklet/interface/PurgeDuplicate.h]: done, and name changed

File: L1Trigger/TrackFindingTracklet/src/PurgeDuplicate.cc (hunk 3 above)
Comment (sarafiorendi) [reply L1Trigger/TrackFindingTracklet/src/PurgeDuplicate.cc]: updated

File: L1Trigger/TrackFindingTracklet/src/PurgeDuplicate.cc (hunk 4 above)
Comment (sarafiorendi) [reply L1Trigger/TrackFindingTracklet/src/PurgeDuplicate.cc]: done

File: L1Trigger/TrackFindingTracklet/python/Tracklet_cfi.py (hunk 5 above)
Comment (sarafiorendi) [reply L1Trigger/TrackFindingTracklet/python/Tracklet_cfi.py]: yes so that it can be changed w/o compiling the code

Comment (tomalin) [reply L1Trigger/TrackFindingTracklet/python/Tracklet_cfi.py]: In that case, I suggest removing it from Tracklet_cfi.py. The same argument would apply to every other parameter in Settings.h. The problem with putting params in Tracklet_cfi.py is that the value specified in Settings.h is then ignored, which tends to confuse people. Ahd we won't change this parameter of yours very often.
(The use of Settings.h is ugly, and forced on us by the request of some members of L1TrkAlgo that the code can also run outside CMSSW. I hope we can move away from this in the Future emulation.)

**Comment by tomalin** [comment]
I suggest setting inventStubs_=true, since whilst it is false, we are simulating an algo that we don't know how to implement in FW.

**Review by tomalin — APPROVED** [review]
Looks good

**Pull request merged**